File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValid( Long.MIN_VALUE , root, Long.MAX_VALUE );
    }

    public boolean isValid(long minval , TreeNode root, long maxval) {
        if(root == null) return true;

        if(root.val >= maxval || root.val <= minval) return false;

        return  isValid(minval, root.left , root.val ) && 
                isValid(root.val , root.right, maxval);

    }

}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValid( LONG_MIN , root, LONG_MAX);
    }

    bool isValid(long minval, TreeNode* root, long maxval) {
        if(root == nullptr) return true;

        if(root->val >= maxval || root->val <= minval) return false;

        return isValid(minval, root->left , root->val) && 
                isValid(root->val, root->right, maxval);

    }
};